riscv/src/continuations/memory_merkle_tree.rs

```diff
@@ -0,0 +1,272 @@
+use std::collections::BTreeMap;
+
+use crate::bootloader::{
+    BYTES_PER_WORD, MERKLE_TREE_DEPTH as MERKLE_TREE_DEPTH_BOOTLOADER,
+    WORDS_PER_PAGE as WORDS_PER_PAGE_BOOTLOADER,
+};
+use number::FieldElement;
+use riscv_executor::poseidon_gl::poseidon_gl;
+
+/// A Merkle tree of memory pages.
+pub struct MerkleTree<
+    T: FieldElement,
+    const MERKLE_TREE_DEPTH: usize = MERKLE_TREE_DEPTH_BOOTLOADER,
+    const WORDS_PER_PAGE: usize = WORDS_PER_PAGE_BOOTLOADER,
+> {
+    /// Hashes of all nodes of the Merkle tree. The vector has MERKLE_TREE_DEPTH + 1 entries
+    /// where hashes[0] is [root_hash], hashes[1] is [level_1_hash_0, level_1_hash_1], etc.
```

Comment:
So the hashes are sorted in breadth-first-order starting from the root, right?

Reply:
Wait, why is it a vector of vectors then? Also why is an entry always four field elements, is that because of poseidon or does it make assumptions about the field size?

Reply:
Oh the first dimension is the level?

Reply:
Right, maybe I should have read the comment more carefully.

Reply:
LOL